s3fs/core.py

```diff
@@ -1129,12 +1136,24 @@ async def _info(self, path, bucket=None, key=None, refresh=False, version_id=Non
         if path in ["/", ""]:
             return {"name": path, "size": 0, "type": "directory"}
         version_id = _coalesce_version_id(path_version_id, version_id)
-        if not refresh and self._ls_from_cache(path) is not None:
+        if not refresh:
             out = self._ls_from_cache(path)
-            out = [o for o in out if o["name"] == path]
-            if out:
-                return out[0]
-            return {"name": path, "size": 0, "type": "directory"}
+            if out is not None:
+                if self.version_aware and version_id is not None:
+                    # If cached info does not match requested version_id,
+                    # fallback to calling head_object
```

Comment:
So looks like it is kinda defeating the purpose of dircache, right? Since calling head_object for every object from dircache is more expensive than just dropping the dircache and doing a list_object_versions call.